[processor/tailsampling] record sampling policy

djluck opened the pull request:
#### Description
Adds support for optionally recording the policy (and any composite policy) associated with an inclusive tail processor sampling decision:

![image](https://github.com/user-attachments/assets/29b5ac62-23fa-4226-9130-1717c1023319)


#### Link to tracking issue
Fixes
Resolves #35180.

#### Implementation notes
- This functionality lives behind a feature flag that is disabled by default
- The original issue described a solution where we might attach the attribute solely to the root span. I'm not sure I agree with the commenter that we can rely on this (e.g. we might decide to sample halfway through a long-running trace) so I have attached the attributes to all present scope spans. This feels like a decent trade off between complexity + network cost, as finding the highest non-root parent would require multiple passes of the spans and keeping all span ids in a set

#### Testing
- Added automated tests to verify enabling the flag both records the expected decision while not impacting existing logic
- Built a custom version and ran it in our preprod environment to ensure it was stable over a 1h period (still evaluating, will update PR with any further observations)

#### TODO
Does this require a CHANGELOG entry?



<!--Please delete paragraphs that you did not use before submitting.-->

djluck committed:
Adds support for optionally recording the policy (and any composite policy) associated with an inclusive tail processor sampling decision.

Resolves !35180.

- This functionality lives behind a feature flag that is disabled by default
- The original issue described a solution where we might attach the attribute solely to the root span. I'm not sure I agree with the commenter that we can rely on this (e.g. we might decide to sample halfway through a long-running trace) so I have attached the attributes to all present scope spans. This feels like a decent trade off between complexity + network cost, as finding the highest non-root parent would require multiple passes of the spans and keeping all span ids in a set

- Added automated tests to verify enabling the flag both records the expected decision while not impacting existing logic
- Built a custom version and ran it in our preprod environment to ensure it was stable over a 1h period (still evaluating, will update PR with any further observations)

Does this require a CHANGELOG entry?

djluck committed:
Adds support for optionally recording the policy (and any composite policy) associated with an inclusive tail processor sampling decision.

Resolves !35180.

- This functionality lives behind a feature flag that is disabled by default
- The original issue described a solution where we might attach the attribute solely to the root span. I'm not sure I agree with the commenter that we can rely on this (e.g. we might decide to sample halfway through a long-running trace) so I have attached the attributes to all present scope spans. This feels like a decent trade off between complexity + network cost, as finding the highest non-root parent would require multiple passes of the spans and keeping all span ids in a set

- Added automated tests to verify enabling the flag both records the expected decision while not impacting existing logic
- Built a custom version and ran it in our preprod environment to ensure it was stable over a 1h period (still evaluating, will update PR with any further observations)

Does this require a CHANGELOG entry?

djluck commented:
Hey @djaglowski, sorry to bother you but do you think you'd be able to take a look at this some time soon?

jpkrohling reviewed (dismissed):
I still prefer to set the policy at the root span, but I acknowledge that the root span might not be easy to find given the current structure we have. I think we can go with this first implementation and see what real-world effects it has on people.

jpkrohling commented on processor/tailsamplingprocessor/processor_decisions_test.go:
```diff
@@ -5,6 +5,7 @@ package tailsamplingprocessor
 
 import (
 	"context"
+	"github.com/stretchr/testify/assert"
```
This will likely not pass the linter, I'd recommend running `make lint` to catch formatting issues before the CI does it.

jpkrohling commented on processor/tailsamplingprocessor/internal/sampling/util.go:
```diff
@@ -93,3 +93,16 @@ func invertHasInstrumentationLibrarySpanWithCondition(ilss ptrace.ScopeSpansSlic
 	return true
 }
+
+func SetAttrOnScopeSpans(data *TraceData, attrName string, attrKey string) {
```
It would be nice to have a test for this function here, and perhaps a benchmark. This way, we can compare it in the future with implementations that find the root span and apply it to that only.

jpkrohling reviewed (commented):
I like this change, it goes in the right direction, but it's missing _unit_ tests for the newly introduced function, a changelog entry, and something in the README telling people how to use this.

djluck commented:
Thanks for your feedback @jpkrohling, I'll work on resolving these comments over the next couple of days 👍

djluck committed:
- Added changelog

- Added README entry for the feature flag
- Added missing mutex lock around reading trace data in `SetAttrOnScopeSpans`
- Added tests + benchmarks for `SetAttrOnScopeSpans`

djluck committed:
Running `make fmt`

djluck committed:
Running `make fmt`

djluck committed:
Fixing `make lint` errors

djluck commented:
@jpkrohling thanks for the review, I've pushed fixes for all your comments. I think I missed `Lock`ing the tracedata mutex so I added this additionally.

Unsure why the pipeline is failing, seems unrelated to anything I touched (command missing error).

jpkrohling commented on processor/tailsamplingprocessor/README.md:
```diff
@@ -528,6 +528,16 @@ sum (otelcol_processor_tail_sampling_count_traces_sampled) by (policy)
+### Tracking sampling policy
+To better understand _which_ sampling policy made the decision to include a trace, you can enable tracking the policy responsible for sampling a trace via the `processor.tailsamplingprocessor.recordpolicy` feature gate.
+
+When this feature gate is set, this will add additional attributes on each sampled span scope:
```
```suggestion
When this feature gate is set, this will add additional attributes on each sampled span:
```

jpkrohling commented on processor/tailsamplingprocessor/internal/sampling/composite_test.go:
```diff
@@ -89,13 +89,34 @@ func TestCompositeEvaluatorSampled(t *testing.T) {
+	expected := Sampled
+	assert.Equal(t, expected, decision)
+	val, ok := trace.ReceivedBatches.ResourceSpans().At(0).ScopeSpans().At(0).Scope().Attributes().Get("tailsampling.composite_policy")
+	if !ok {
```
You can use an `assert.True(t, ok, "Did not ...")` call instead

jpkrohling commented on processor/tailsamplingprocessor/internal/sampling/composite_test.go:
```diff
@@ -160,12 +181,35 @@ func TestCompositeEvaluatorInverseSampled_AlwaysSampled(t *testing.T) {
+		expected := Sampled
+		assert.Equal(t, expected, decision)
+		val, ok := trace.ReceivedBatches.ResourceSpans().At(0).ScopeSpans().At(0).Scope().Attributes().Get("tailsampling.composite_policy")
+		if !ok {
```
same as above: `assert.True` is preferable

jpkrohling reviewed (approved):
I'm OK merging as is, the comments are nits.

jpkrohling commented:
The code owners check seems to be failing for all jobs at the moment.

djluck committed:
Making suggested fixes

djluck commented:
@jpkrohling thanks for the review, I've pushed fixes for your comments

jpkrohling commented:
Linting seems to be failing:

```
tailsamplingprocessor/factory.go:12: File is not `gci`-ed with --skip-generated -s standard -s default -s prefix(github.com/open-telemetry/opentelemetry-collector-contrib) (gci)
	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor/internal/telemetry"

tailsamplingprocessor/factory.go:18: File is not `gci`-ed with --skip-generated -s standard -s default -s prefix(github.com/open-telemetry/opentelemetry-collector-contrib) (gci)
	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor/internal/metadata"
tailsamplingprocessor/processor_decisions_test.go:12: File is not `gci`-ed with --skip-generated -s standard -s default -s prefix(github.com/open-telemetry/opentelemetry-collector-contrib) (gci)
```

You can run `make lint` locally to test the changes before going to the CI.

djluck committed:
Running `make gci`

djluck commented:
@jpkrohling apologies, re-ran `make lint`, `make fmt` + `make gci`- I think we should be good now

djluck commented:
Thanks again @jpkrohling for reviewing this change but I think I need to bother you one last time: I can see you've approved but I'm not authorized to merge. Is that something you could do?

djluck commented:
Hey @jpkrohling! Just a friendly ping- if you've got time before Christmas and you're able to help merge this it would be appreciated. Otherwise I can follow up in the new year.

djluck commented:
@jpkrohling I've rebased, reformatted and pushed. Do you mind reopening this MR and merging this change?

purkhusid commented:
Let's get this merged, this is a very useful feature!

djluck commented:
Hey @jpkrohling, a friendly nudge about reopening this MR when you get the time

djluck commented:
Reopened as a new PR: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37797. Apologies if this was the wrong process, was unsure on the best way to get this reviewed.

jpkrohling commented:
I'm sorry for dropping the ball on this one, I'll review the new PR right now.
